pyneuroml/neuron/__init__.py

```diff
@@ -403,6 +406,7 @@ def cellinfo(doprint: str = "") -> dict:
                 for j in range(numParams):
                     ms.name(pname, j)
                     totParamVal[j] += ms.get(pname)
+                    paramsectiondict[pname[0]][str(sec)] = ms.get(pname)
```

Comment:
I think we should add an extra loop iterating on # nseg, printing the gmax for each of the segments. But we can limit this just to gmax as all the other parameters are set by nml file and are fixed for a section across segments.
@sanjayankur31 
Right now, its printing a single value even for the nonhomogenous density mechanisms having nseg more than 1.

Reply:
Hrm, yeh, at the moment this is only at the section level, not at the segment level. 

 These utils aren't limited to neuroml generated neuron, they should work even with hand written neuron files (in fact one of their uses is to compare the hand written neuron files with neuroml generated ones to see if we're getting the same thing). So, will `gmax` be the only thing that changes in segments, can other parameters not change? I think technically other parameters can also change: https://docs.neuroml.org/Userdocs/Schemas/Cells.html?highlight=inhomo#variableparameter

Reply:
But how do we run it for hand written neuron files? I tried running it for [this](https://github.com/OpenSourceBrain/BahlEtAl2012_ReducedL5PyrCell/blob/master/NEURON/test/Test_example1.hoc) as it wasnt able to access the cell correctly and hence the `seclist` as defined in the `cellinfo()`. Running cellinfo() after loading the `Test_example1.hoc` file gave the following error:
<img width="711" alt="image" src="https://user-images.githubusercontent.com/16141677/186460378-13c7421d-a95e-496b-96ab-0edcffdba152.png">

Reply:
let me test that out. (I've been running it on the Hay et al NEURON files and it's worked fine there.)

Reply:
hrm, I'm looking at the NEURON files. Looks like the sections have not been linked to a cell at all, there are only sections connected to each other? So, `secinfo()` works fine, but that doesn't have all the information that we've put into `cellinfo()`.

Let me think of a way of refactoring this so that we can maybe also provide a section list to go over in addition to being able to get it from a cell.

Reply:
Okay but the hoc file I mentioned has other sections [not just soma] as well. So there should be other sections as well in the list while passing that to `seclistinfo()`. You can try running `secinfo` on tuft by using secinfo('tuft') and it will show the desired results.

For running `seclistinfo` on sections other than soma, I tried this `seclistinfo([h.tuft, h.apical, h.soma])` for [this hoc file](https://github.com/OpenSourceBrain/BahlEtAl2012_ReducedL5PyrCell/blob/master/NEURON/test/Test_example1.hoc). It returns the info related to all the three sections tuft, apical and soma at once like this:

```
ca_ion:
    avg_rev_pot: 140.0
    ext_conc: 2.0
    int_conc: 0.00011085902659802099
    nsecs: 1
k_ion:
    avg_rev_pot: -80.0
    ext_conc: 2.5
    int_conc: 54.4
    nsecs: 3
mechanisms:
    ca_ion:
        num_params: 1
        parameters:
            cao:
                ave_all_sections: 2.0
                values:
                    tuft: 2.0
        present_on: 1
    cad:
        num_params: 0
        parameters: {}
        present_on: 1
    capacitance:
        num_params: 1
        parameters:
            cm:
                ave_all_sections: 1.9104143381173333
                values:
                    apical: 1.750417507176
                    soma: 2.230408
                    tuft: 1.750417507176
        present_on: 3
    extracellular:
        num_params: 4
        parameters:
            e_extracellular:
                ave_all_sections: NA
                values: NA
            xc:
                ave_all_sections: NA
                values: NA
            xg:
                ave_all_sections: NA
                values: NA
            xraxial:
                ave_all_sections: NA
                values: NA
        present_on: 0
    fastpas:
        num_params: 2
        parameters:
            e_fastpas:
                ave_all_sections: NA
                values: NA
            g_fastpas:
                ave_all_sections: NA
                values: NA
        present_on: 0
    hh:
        num_params: 4
        parameters:
            el_hh:
                ave_all_sections: NA
                values: NA
            gkbar_hh:
                ave_all_sections: NA
                values: NA
            gl_hh:
                ave_all_sections: NA
                values: NA
            gnabar_hh:
                ave_all_sections: NA
                values: NA
        present_on: 0
    ih:
        num_params: 3
        parameters:
            ehd_ih:
                ave_all_sections: -47.0
                values:
                    apical: -47.0
                    tuft: -47.0
            gbar_ih:
                ave_all_sections: 9.229572880315631
                values:
                    apical: 2.2643307606312626
                    tuft: 16.194815
            vshift_ih:
                ave_all_sections: 0.0
                values:
                    apical: 0.0
                    tuft: 0.0
        present_on: 2
    k_ion:
        num_params: 1
        parameters:
            ko:
                ave_all_sections: 2.5
                values:
                    apical: 2.5
                    soma: 2.5
                    tuft: 2.5
        present_on: 3
    kca:
        num_params: 1
        parameters:
            gbar_kca:
                ave_all_sections: 6.15058501
                values:
                    tuft: 6.15058501
        present_on: 1
    kfast:
        num_params: 3
        parameters:
            gbar_kfast:
                ave_all_sections: 22.9851401757044
                values:
                    apical: 1.7579125271125555
                    soma: 67.197508
                    tuft: 6.47730835627545e-13
            timefactor_n_kfast:
                ave_all_sections: 1.0
                values:
                    apical: 1.0
                    soma: 1.0
                    tuft: 1.0
            vshift_kfast:
                ave_all_sections: 0.0
                values:
                    apical: 0.0
                    soma: 0.0
                    tuft: 0.0
        present_on: 3
    km:
        num_params: 2
        parameters:
            gbar_km:
                ave_all_sections: 10.459916
                values:
                    soma: 10.459916
            vshift_km:
                ave_all_sections: 0.0
                values:
                    soma: 0.0
        present_on: 1
    kslow:
        num_params: 3
        parameters:
            gbar_kslow:
                ave_all_sections: 181.40879466693085
                values:
                    apical: 68.40572154979573
                    soma: 475.820646
                    tuft: 1.645099680002128e-05
            vshift2_kslow:
                ave_all_sections: 0.0
                values:
                    apical: 0.0
                    soma: 0.0
                    tuft: 0.0
            vshift_kslow:
                ave_all_sections: 0.0
                values:
                    apical: 0.0
                    soma: 0.0
                    tuft: 0.0
        present_on: 3
    morphology:
        num_params: 1
        parameters:
            diam:
                ave_all_sections: 11.697256724039713
                values:
                    apical: 5.928448677062988
                    soma: 23.14525604248047
                    tuft: 6.018065452575683
        present_on: 3
    na_ion:
        num_params: 1
        parameters:
            nao:
                ave_all_sections: 140.0
                values:
                    apical: 140.0
                    soma: 140.0
                    tuft: 140.0
        present_on: 3
    nap:
        num_params: 2
        parameters:
            gbar_nap:
                ave_all_sections: 1.443953
                values:
                    soma: 1.443953
            vshift_nap:
                ave_all_sections: 0.0
                values:
                    soma: 0.0
        present_on: 1
    nat:
        num_params: 6
        parameters:
            gbar_nat:
                ave_all_sections: 164.88424584715395
                values:
                    apical: 210.21872154146186
                    soma: 236.616175
                    tuft: 47.817841
            gbarfactor_nat:
                ave_all_sections: 1.0
                values:
                    apical: 1.0
                    soma: 1.0
                    tuft: 1.0
            timefactor_h_nat:
                ave_all_sections: 1.0
                values:
                    apical: 1.0
                    soma: 1.0
                    tuft: 1.0
            timefactor_m_nat:
                ave_all_sections: 1.0
                values:
                    apical: 1.0
                    soma: 1.0
                    tuft: 1.0
            vshift2_nat:
                ave_all_sections: 0.0
                values:
                    apical: 0.0
                    soma: 0.0
                    tuft: 0.0
            vshift_nat:
                ave_all_sections: 10.0
                values:
                    apical: 10.0
                    soma: 10.0
                    tuft: 10.0
        present_on: 3
    pas:
        num_params: 2
        parameters:
            e_pas:
                ave_all_sections: -80.398657
                values:
                    apical: -80.398657
                    soma: -80.398657
                    tuft: -80.398657
            g_pas:
                ave_all_sections: 4.1603962264820595e-05
                values:
                    apical: 3.811963848009967e-05
                    soma: 4.857260983426245e-05
                    tuft: 3.811963848009967e-05
        present_on: 3
    sca:
        num_params: 3
        parameters:
            actF_sca:
                ave_all_sections: 1.0
                values:
                    tuft: 1.0
            gbar_sca:
                ave_all_sections: 0.45423528
                values:
                    tuft: 0.45423528
            inactF_sca:
                ave_all_sections: 3.0
                values:
                    tuft: 3.0
        present_on: 1
na_ion:
    avg_rev_pot: 55.0
    ext_conc: 140.0
    int_conc: 10.0
    nsecs: 3
num_sections: 3
temperature: 37.0
total_Cm: 5.731243014352
total_L: 1022.1452564588122
total_Ra: 991.22414867
total_diam: 35.09177017211914
total_nseg: 8
```

Reply:
Yeh, one can pass as many sections as required to the function. Good that
it works. We'll merge this now and keep adding more bits as required.

Reply:
@shayan823 : check now, it provides information by each segment (beware: output is much larger).

Also note that I've renamed some functions. named it `getinfo` now---got a better suggestion?

Reply:
I tried that on the section `apical` which has `nonhomogenous nat, ih` channel. But it throws out just one value. 

I think we can name the function `secinfo() `and rename the existing file of the same name with `secsummary()`.

Here is the output when I ran it for Bahl cell on section `apical`
```
>>> getinfo([h.apical],"yaml")
ca_ion:
    avg_rev_pot: NA
    ext_conc: NA
    int_conc: NA
    nsecs: 0
k_ion:
    avg_rev_pot: -80.0
    ext_conc: 2.5
    int_conc: 54.4
    nsecs: 1
mechanisms:
    ca_ion:
        num_params: 0
        parameters: {}
        present_on_secs: 0
    cad:
        num_params: 0
        parameters: {}
        present_on_secs: 0
    capacitance:
        num_params: 1
        parameters:
            cm:
                ave_all_sections: 8.75208753588
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 1.750417507176
        present_on_secs: 1
    extracellular:
        num_params: 4
        parameters:
            e_extracellular:
                ave_all_sections: NA
                values: NA
            xc:
                ave_all_sections: NA
                values: NA
            xg:
                ave_all_sections: NA
                values: NA
            xraxial:
                ave_all_sections: NA
                values: NA
        present_on_secs: 0
    fastpas:
        num_params: 2
        parameters:
            e_fastpas:
                ave_all_sections: NA
                values: NA
            g_fastpas:
                ave_all_sections: NA
                values: NA
        present_on_secs: 0
    hh:
        num_params: 4
        parameters:
            el_hh:
                ave_all_sections: NA
                values: NA
            gkbar_hh:
                ave_all_sections: NA
                values: NA
            gl_hh:
                ave_all_sections: NA
                values: NA
            gnabar_hh:
                ave_all_sections: NA
                values: NA
        present_on_secs: 0
    ih:
        num_params: 3
        parameters:
            ehd_ih:
                ave_all_sections: -235.0
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 0.0
            gbar_ih:
                ave_all_sections: 43.826117028350026
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 0.0
            vshift_ih:
                ave_all_sections: 0.0
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 0.0
        present_on_secs: 1
    k_ion:
        num_params: 1
        parameters:
            ko:
                ave_all_sections: 12.5
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 2.5
        present_on_secs: 1
    kca:
        num_params: 1
        parameters:
            gbar_kca:
                ave_all_sections: NA
                values: NA
        present_on_secs: 0
    kfast:
        num_params: 3
        parameters:
            gbar_kfast:
                ave_all_sections: 1.770065512042127
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 1.0
            timefactor_n_kfast:
                ave_all_sections: 5.0
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 1.0
            vshift_kfast:
                ave_all_sections: 0.0
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 1.0
        present_on_secs: 1
    km:
        num_params: 2
        parameters:
            gbar_km:
                ave_all_sections: NA
                values: NA
            vshift_km:
                ave_all_sections: NA
                values: NA
        present_on_secs: 0
    kslow:
        num_params: 3
        parameters:
            gbar_kslow:
                ave_all_sections: 73.59526820903713
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 0.0
            vshift2_kslow:
                ave_all_sections: 0.0
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 0.0
            vshift_kslow:
                ave_all_sections: 0.0
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 0.0
        present_on_secs: 1
    morphology:
        num_params: 1
        parameters:
            diam:
                ave_all_sections: 29.64224338531494
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 5.928448677062988
        present_on_secs: 1
    na_ion:
        num_params: 1
        parameters:
            nao:
                ave_all_sections: 700.0
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 140.0
        present_on_secs: 1
    nap:
        num_params: 2
        parameters:
            gbar_nap:
                ave_all_sections: NA
                values: NA
            vshift_nap:
                ave_all_sections: NA
                values: NA
        present_on_secs: 0
    nat:
        num_params: 6
        parameters:
            gbar_nat:
                ave_all_sections: 672.1582197772316
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 1.0
            gbarfactor_nat:
                ave_all_sections: 5.0
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 1.0
            timefactor_h_nat:
                ave_all_sections: 5.0
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 1.0
            timefactor_m_nat:
                ave_all_sections: 5.0
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 1.0
            vshift2_nat:
                ave_all_sections: 0.0
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 1.0
            vshift_nat:
                ave_all_sections: 50.0
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': 1.0
        present_on_secs: 1
    pas:
        num_params: 2
        parameters:
            e_pas:
                ave_all_sections: -401.993285
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': -80.398657
            g_pas:
                ave_all_sections: 0.00019059819240049833
                values:
                    apical:
                        nseg: 5
                        values:
                            '*': -80.398657
        present_on_secs: 1
    sca:
        num_params: 3
        parameters:
            actF_sca:
                ave_all_sections: NA
                values: NA
            gbar_sca:
                ave_all_sections: NA
                values: NA
            inactF_sca:
                ave_all_sections: NA
                values: NA
        present_on_secs: 0
na_ion:
    avg_rev_pot: 55.0
    ext_conc: 140.0
    int_conc: 10.0
    nsecs: 1
num_sections: 1
temperature: 37.0
total_Cm: 1.750417507176
total_L: 500.0
total_Ra: 382.22414867
total_diam: 5.928448677062988
total_nseg: 5
```

Reply:
Check my comment below---if it's showing one value, that implies that values on all segments are the same. Is that not the case in this model?

Reply:
No, it isnt. I will check running the same on neuroml generated neuron file

Reply:
I still dont know why is this throwing the same `gmax` values for ih, nat channel with both being `nonhomogenous`.

Reply:
how does this look now?

```

    nat:
        num_params: 6
        parameters:
            gbar_nat:
                ave_all_segs: 134.4316439554463
                values:
                    apical:
                        id: apical
                        nseg: 5
                        values:
                            1: 210.21872154146186
                            2: 174.12963697669255
                            3: 138.04055241192324
                            4: 101.95146784715394
                            5: 47.81784099999999
            gbarfactor_nat:
                ave_all_segs: 1.0
                values:
                    apical:
                        id: apical
                        nseg: 5
                        values:
                            '*': 1.0
            timefactor_h_nat:
                ave_all_segs: 1.0
                values:
                    apical:
                        id: apical
                        nseg: 5
                        values:
                            '*': 1.0
            timefactor_m_nat:
                ave_all_segs: 1.0
                values:
                    apical:
                        id: apical
                        nseg: 5
                        values:
                            '*': 1.0
            vshift2_nat:
                ave_all_segs: 0.0
                values:
                    apical:
                        id: apical
                        nseg: 5
                        values:
                            '*': 0.0
            vshift_nat:
                ave_all_segs: 10.0
                values:
                    apical:
                        id: apical
                        nseg: 5
                        values:
                            '*': 10.0
        present_on_secs: 1


```

Reply:
Expected the values as it is here. Looks okay to me.